src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -81,72 +80,87 @@ private static ulong GetMaskAsLong(ReadOnlySpan<byte> bytes)
         if (span.IsEmpty)
         {
             return string.Empty;
         }
 
-        fixed (byte* source = &MemoryMarshal.GetReference(span))
+        var str = string.Create(span.Length, span, static (destination, source) =>
         {
-            return string.Create(span.Length, new IntPtr(source), s_getHeaderName);
-        }
-    }
-
-    private static unsafe void GetHeaderName(Span<char> buffer, IntPtr state)
-    {
-        fixed (char* output = &MemoryMarshal.GetReference(buffer))
-        {
-            // This version of AsciiUtilities returns null if there are any null (0 byte) characters
-            // in the string
-            if (!StringUtilities.TryGetAsciiString((byte*)state.ToPointer(), output, buffer.Length))
+            if (Ascii.ToUtf16(source, destination, out var written) != OperationStatus.Done
+                || destination.Contains('\0'))
             {
                 KestrelBadHttpRequestException.Throw(RequestRejectionReason.InvalidCharactersInHeaderName);
             }
-        }
+
+            Debug.Assert(written == destination.Length);
+        });
+
+        return str;
     }
 
-    public static string GetAsciiStringNonNullCharacters(this Span<byte> span)
-        => StringUtilities.GetAsciiStringNonNullCharacters(span);
+    // Null checks must be done independently of this method (if required)
+    public static string GetAsciiString(this Span<byte> span)
+        => StringUtilities.GetAsciiString(span);
 
-    public static string GetAsciiOrUTF8StringNonNullCharacters(this ReadOnlySpan<byte> span)
-        => StringUtilities.GetAsciiOrUTF8StringNonNullCharacters(span, DefaultRequestHeaderEncoding);
+    // Null checks must be done independently of this method (if required)
+    public static string GetAsciiOrUTF8String(this ReadOnlySpan<byte> span)
+        => StringUtilities.GetAsciiOrUTF8String(span, DefaultRequestHeaderEncoding);
 
     public static string GetRequestHeaderString(this ReadOnlySpan<byte> span, string name, Func<string, Encoding?> encodingSelector, bool checkForNewlineChars)
     {
         string result;
         if (ReferenceEquals(KestrelServerOptions.DefaultHeaderEncodingSelector, encodingSelector))
         {
-            result = span.GetAsciiOrUTF8StringNonNullCharacters(DefaultRequestHeaderEncoding);
+            result = span.GetAsciiOrUTF8String(DefaultRequestHeaderEncoding);
         }
         else
         {
             result = span.GetRequestHeaderStringWithoutDefaultEncodingCore(name, encodingSelector);
         }
 
         // New Line characters (CR, LF) are considered invalid at this point.
-        if (checkForNewlineChars && ((ReadOnlySpan<char>)result).IndexOfAny('\r', '\n') >= 0)
+        // Null characters are also not allowed.
+        var invalidCharIndex = checkForNewlineChars ?
+            ((ReadOnlySpan<char>)result).IndexOfAny('\r', '\n', '\0')
+            : ((ReadOnlySpan<char>)result).IndexOf('\0');
+
+        if (invalidCharIndex >= 0)
```

Comment:
```suggestion
        var invalidChar = checkForNewlineChars ?
            ((ReadOnlySpan<char>)result).ContainsAny('\r', '\n', '\0')
            : ((ReadOnlySpan<char>)result).Contains('\0');

        if (invalidChar)
```

Contains is faster as IndexOf is Contains+Resolving the postive hit to an exact index